### PR TITLE
checker: refactor `fixRulePeer`

### PR DESCRIPTION
**Pull request by rleungx**
Signed-off-by: Ryan Leung <rleungx@gmail.com>

### What problem does this PR solve?

Issue Number: Close #xxx

### What is changed and how does it work?

```commit-message
```

### Check List

Tests

- Unit test

### Release note

```release-note
None
```

### CODE REVIEW

File: server/schedule/checker/rule_checker.go [hunk 1]
```diff
@@ -135,22 +131,64 @@ func (c *RuleChecker) CheckWithFit(region *core.RegionInfo, fit *placement.Regio
...
-		if c.isOfflinePeer(peer) {
-			checkerCounter.WithLabelValues("rule_checker", "replace-offline").Inc()
-			return c.replaceUnexpectRulePeer(region, rf, fit, peer, offlineStatus)
+
+		// check orphan peer.
+		if !skipOrphan {
+			for _, pendingPeer := range region.GetPendingPeers() {
+				if pendingPeer.Id == peer.Id {
+					checkerCounter.WithLabelValues("rule_checker", "skip-remove-pending-peer").Inc()
+					skipOrphan = true
```
Comment (AndreMouche) [inline server/schedule/checker/rule_checker.go:163]: could we break here?

File: server/schedule/checker/rule_checker.go [hunk 2]
```diff
@@ -135,22 +131,64 @@ func (c *RuleChecker) CheckWithFit(region *core.RegionInfo, fit *placement.Regio
...
+				}
+			}
+			for _, downPeer := range region.GetDownPeers() {
```
Comment (AndreMouche) [inline server/schedule/checker/rule_checker.go:166]: should we check `skipOrphan` before loop?

File: server/schedule/checker/rule_checker.go [hunk 3]
```diff
@@ -135,22 +131,64 @@ func (c *RuleChecker) CheckWithFit(region *core.RegionInfo, fit *placement.Regio
...
+				if downPeer.Peer.Id == peer.Id {
+					checkerCounter.WithLabelValues("rule_checker", "skip-remove-down-peer").Inc()
+					skipOrphan = true
```
Comment (AndreMouche) [inline server/schedule/checker/rule_checker.go:169]: ditto, could we break here?

File: server/schedule/checker/rule_checker.go [hunk 4]
```diff
@@ -135,22 +131,64 @@ func (c *RuleChecker) CheckWithFit(region *core.RegionInfo, fit *placement.Regio
...
+		if selectedPeer == nil && c.isOfflinePeer(peer) {
+			selectedPeer, kind = peer, offlineStatus
 		}
-		if c.isOfflinePeer(peer) {
-			checkerCounter.WithLabelValues("rule_checker", "replace-offline").Inc()
-			return c.replaceUnexpectRulePeer(region, rf, fit, peer, offlineStatus)
+
+		// check orphan peer.
+		if !skipOrphan {
+			for _, pendingPeer := range region.GetPendingPeers() {
```
Comment (AndreMouche) [inline server/schedule/checker/rule_checker.go:160]: how about define some function like  containsPeer(peers,peerID) bool?
if it is only used in this func, define it as a temp variable should be ok, for example,
```
	constainsPeer := func(peers []*pdpb.PeerStats, peerID uint64) bool {
		for _, peer := range peers {
			if peer.Peer.Id == peerID {
				return true
			}
		}
		return false
	}
```

File: server/schedule/checker/rule_checker.go [hunk 5]
```diff
@@ -135,22 +131,64 @@ func (c *RuleChecker) CheckWithFit(region *core.RegionInfo, fit *placement.Regio
...
+		}
+		if selectedPeer == nil && c.isOfflinePeer(peer) {
+			selectedPeer, kind = peer, offlineStatus
 		}
-		if c.isOfflinePeer(peer) {
-			checkerCounter.WithLabelValues("rule_checker", "replace-offline").Inc()
-			return c.replaceUnexpectRulePeer(region, rf, fit, peer, offlineStatus)
+
+		// check orphan peer.
+		if !skipOrphan {
```
Comment (AndreMouche) [inline server/schedule/checker/rule_checker.go:159]: Too many {} in {}, could we continue directly when skipOrphan is true?

File: server/schedule/checker/rule_checker.go [hunk 6]
```diff
@@ -107,15 +107,10 @@ func (c *RuleChecker) CheckWithFit(region *core.RegionInfo, fit *placement.Regio
...
-		log.Debug("fail to fix orphan peer", errs.ZapError(err))
-	} else if op != nil {
-		c.pendingList.Remove(region.GetID())
-		return op
-	}
+
+	skip := len(fit.OrphanPeers) == 0
 	for _, rf := range fit.RuleFits {
-		op, err := c.fixRulePeer(region, fit, rf)
+		op, err := c.fixRulePeer(region, fit, rf, skip)
```
[the remark this answers is not in the document]
Comment (rleungx) [reply server/schedule/checker/rule_checker.go:113]: It's unnecessary to check it every loop.

File: server/schedule/checker/rule_checker.go (hunk 5 above)
Comment (rleungx) [reply server/schedule/checker/rule_checker.go:159]: What do you mean `continue`?

File: server/schedule/checker/rule_checker.go (hunk 1 above)
Comment (rleungx) [reply server/schedule/checker/rule_checker.go:163]: No, we can't.

File: server/schedule/checker/rule_checker.go (hunk 4 above)
Comment (rleungx) [reply server/schedule/checker/rule_checker.go:160]: The down peer and pending peer are stored in the different struct, which cannot share the same function.

**Comment by rleungx** [comment]
Close it for now.

**Pull request closed without merging**